src/runtime/vm/debug/debug_vm.cc

```diff
@@ -0,0 +1,108 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file src/runtime/vm/debug/vm_debug.cc
+ * \brief The Relay debug virtual machine.
+ */
+
+#include <tvm/runtime/registry.h>
+#include <tvm/runtime/vm.h>
+
+#include <chrono>
+#include <iomanip>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "debug_vm.h"
+
+namespace tvm {
+namespace runtime {
+namespace vm {
+
+PackedFunc VirtualMachineDebug::GetFunction(
+    const std::string& name, const std::shared_ptr<ModuleNode>& sptr_to_self) {
+  if (name == "get_stat") {
+    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
+      double total_duration = 0.0;
+      std::ostringstream os;
+      os << std::setw(30) << std::left << "#OpName"
+         << "\t" << std::setw(10) << std::left << "#InvokeCount"
+         << "\t"
+         << "#Duration(us)" << std::endl;
+
+      for (auto kv : op_durations) {
+        os << std::setw(30) << std::left << packed_index_map[kv.first] << "\t"
+           << std::setw(10) << std::left << op_invokes[kv.first] << "\t"
+           << op_durations[kv.first] << std::endl;
+
+        total_duration += op_durations[kv.first];
+      }
+      os << "Total Duration " << total_duration << " us" << std::endl;
+      *rv = os.str();
+    });
+  } else if (name == "init") {
+    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
+      CHECK_EQ(args.size() % 2, 0);
+      std::vector<TVMContext> contexts;
+      for (int i = 0; i < args.size() / 2; ++i) {
+        TVMContext ctx;
+        int device_type = args[i * 2];
+        ctx.device_type = DLDeviceType(device_type);
+        ctx.device_id = args[i * 2 + 1];
+        contexts.push_back(ctx);
+      }
+      this->Init(contexts);
+    });
+  } else {
+    return VirtualMachine::GetFunction(name, sptr_to_self);
+  }
+}
+
+void VirtualMachineDebug::Init(const std::vector<TVMContext>& ctxs) {
+  VirtualMachine::Init(ctxs);
+  for (auto kv : primitive_map) {
+    packed_index_map[kv.second] = kv.first;
+    op_durations[kv.second] = 0.0;
+    op_invokes[kv.second] = 0;
+  }
+}
+
+void VirtualMachineDebug::InvokePacked(Index packed_index,
+                                       const PackedFunc& func, Index arg_count,
+                                       Index output_size,
+                                       const std::vector<Object>& args) {
+  auto op_begin = std::chrono::high_resolution_clock::now();
+  VirtualMachine::InvokePacked(packed_index, func, arg_count, output_size,
+                               args);
+  auto op_end = std::chrono::high_resolution_clock::now();
+  double op_duration =
+      std::chrono::duration_cast<std::chrono::duration<double> >(op_end -
```

Comment:
Use `std::chrono::microseconds` instead?

Reply:
I think high_resolution_clock is fine. But do we need to sync before measurement to profile devices other than CPU?

Reply:
maybe I add a todo here once we have heterogeneous execution? https://github.com/dmlc/tvm/blob/master/src/runtime/graph/debug/graph_runtime_debug.cc#L80

Reply:
Since VM now supports running on CUDA, we should probably handle this here.

Reply:
thanks. done.